Add support Flash QSPI on S32Z270

congnguyenhuu opened the pull request:
Introduce support HyperFlash  memory devices on a NXP S32 QSPI bus. This driver uses a fixed LUT configuration that defined in HAL RTD HyperFlash driver and  allows to read, write, and erase HyperFlash devices.
Add support QSPI secure flash protection (SFP) for memory control NXP S32 QSPI.
Enable support Flash QSPI on S32Z2XX, the on-board S26HS512T 512M-bit HyperFlash memory is connected to the QSPI controller port A1. This board configuration selects it as the default flash controller.

The test result:

 ```
west twister  --disable-warnings-as-errors --device-testing --device-serial=/dev/ttyUSB0 -v --runtime-artifact-cleanup --west-runner trace32 --west-flash="--config=./config.t32" -p s32z2xxdc2/s32z270/rtu1  -T samples/subsys/fs/littlefs -T samples/subsys/settings -T samples/subsys/shell/fs -T tests/drivers/flash -T tests/subsys/fs -T tests/subsys/settings -T tests/subsys/storage/flash_map -T tests/subsys/logging/log_backend_fs
INFO    - Using Ninja..
INFO    - Zephyr version: 8c53d91026a8
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...

Device testing on:

| Platform                | ID   | Serial device   |
|-------------------------|------|-----------------|
| s32z2xxdc2/s32z270/rtu1 |      | /dev/ttyUSB0 |

INFO    - JOBS: 6
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 58/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/storage/flash_map/storage.flash_map   PASSED (device 80.480s)
INFO    - 59/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/logging/log_backend_fs/logging.backend.fs.automounted PASSED (device 38.796s)
INFO    - 60/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/file/settings.file PASSED (device 28.017s)
INFO    - 61/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/logging/log_backend_fs/logging.backend.fs.manualmounted PASSED (device 40.086s)
INFO    - 62/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/nvs/settings.functional.nvs.chosen PASSED (device 27.685s)
INFO    - 63/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/fcb/settings.functional.fcb.chosen PASSED (device 27.686s)
INFO    - 64/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/fcb/settings.functional.fcb PASSED (device 27.692s)
INFO    - 65/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/nvs/settings.functional.nvs.dk PASSED (device 27.624s)
INFO    - 66/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/functional/nvs/settings.functional.nvs PASSED (device 27.641s)
INFO    - 67/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/fs/fat_fs_api/filesystem.fat.api.sdmmc SKIPPED (runtime filter)
INFO    - 68/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/fs/fat_fs_api/filesystem.fat.api.mmc  SKIPPED (runtime filter)
INFO    - 69/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/storage/flash_map/storage.flash_map.mbedtls PASSED (device 80.335s)
INFO    - 70/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/storage/flash_map/storage.flash_map.psa PASSED (device 80.570s)
INFO    - 72/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/file/settings.file.raw       PASSED (device 30.420s)
INFO    - 73/77 s32z2xxdc2/s32z270/rtu1   samples/subsys/settings/sample.subsys.settings     PASSED (device 27.357s)
INFO    - 74/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/fs/fs_api/filesystem.api              PASSED (device 26.865s)
INFO    - 75/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/settings/fcb/settings.fcb.raw         PASSED (device 47.201s)
INFO    - 76/77 s32z2xxdc2/s32z270/rtu1   tests/drivers/flash/common/drivers.flash.common.default PASSED (device 26.663s)
INFO    - 77/77 s32z2xxdc2/s32z270/rtu1   tests/subsys/fs/littlefs/filesystem.littlefs.default PASSED (device 39.218s)

INFO    - 77 test scenarios (77 test instances) selected, 59 configurations skipped (57 by static filter, 2 at runtime).
INFO    - 18 of 77 test configurations passed (100.00%), 0 failed, 0 errored, 59 skipped with 0 warnings in 808.67 seconds
INFO    - In total 134 test cases were executed, 355 skipped on 1 out of total 4 platforms (25.00%)
INFO    - 18 test configurations executed on platforms, 0 test configurations were only built.
```

congnguyenhuu committed:
drivers: flash: add NXP S32 QSPI HyperFlash driver

Add support HyperFlash memory devices on a NXP S32 QSPI bus.
This driver uses a fixed LUT configuration that defined in HAL RTD
HyperFlash driver.
Driver allows to read, write and erase HyperFlash devices.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
drivers: memc_nxp_s32_qspi: add support for s32ze

Add support QSPI secure flash protection (SFP)

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
samples: enable flash samples for s32z board

Enable flash samples for s32z board

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
tests: enable flash tests for s32z board

Enable flash tests for s32z board

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
boards: s32z2xx: enable flash controller QSPI

The on-board S26HS512T 512M-bit HyperFlash memory is connected to
the QSPI controller port A1.
This board configuration selects it as the default flash controller.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

Dat-NguyenDuy commented on boards/nxp/s32z2xxdc2/s32z2xxdc2_s32z270_pinctrl.dtsi:
```diff
@@ -107,4 +107,25 @@
+				 <PI2_QSPI_0_CK_A_B>, <PI3_QSPI_0_CK_A>;
+			output-enable;
+		};
+		group5 {
```
why not group 3

congnguyenhuu commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+
+compatible: "nxp,s32-qspi-hyperflash"
+
+include: ["nxp,s32-qspi-device.yaml", "soc-nv-flash.yaml"]
```
I'm using the property `write-block-size` from this

manuargue replied:
this is an external flash, not a soc flash. Also you have write block size hardcoded in https://github.com/zephyrproject-rtos/zephyr/pull/78073/files#diff-aa5c7e912c23e4bbc1cabe0abb767637bcdaf3f9892233d71bd0841fdbec5c93R13 . Move this property to nxp,s32-qspi-device.yaml and unify the behavior for flash nor & hyperflash

congnguyenhuu replied:
I moved it to nxp,s32-qspi-device.yaml and unify the behavior for flash nor & hyperflash

congnguyenhuu commented on drivers/flash/flash_nxp_s32_qspi_hyperflash.c:
```diff
@@ -0,0 +1,527 @@
+		&((const struct nxp_s32_qspi_config *)dev->config)->memory_cfg);
+}
+
+static ALWAYS_INLINE bool area_is_subregion(const struct device *dev, off_t offset, size_t size)
```
yes, there are some duplicated functions.
But I think we should separate code for each shim-driver.
In the future, it is easy for us to convert native driver for each that is not impact to the rest driver.

manuargue replied:
if the intention is to use the HAL, and the HAL uses same API for any kind of memory connected to the QSPI bus, then let's not duplicate code and merge whatever is common for reusability

congnguyenhuu replied:
okay, I updated to create a common source code to use for both

congnguyenhuu committed:
drivers: flash_nxp_s32: create common source code

Create common source code to use for supporting HyperFlash.

Rename 'FLASH_NXP_S32_QSPI_NOR_SFDP_RUNTIME' to
'FLASH_NXP_S32_QSPI_SFDP_RUNTIME' as a common kconfig.

Add the 'max-program-buffer-size' property to use for
setting memory pageSize, instead of using
'CONFIG_FLASH_NXP_S32_QSPI_LAYOUT_PAGE_SIZE' for setting.

Add the 'write-block-size' propertyto use for setting
the number of bytes used in write operations, it also
uses to instead of the 'memory-alignment' property.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

nordicjm commented on drivers/flash/CMakeLists.txt:
```diff
@@ -126,10 +126,13 @@ zephyr_library_include_directories_ifdef(
-  CONFIG_FLASH_NXP_S32_QSPI_NOR
+zephyr_library_sources_ifdef(CONFIG_FLASH_NXP_S32_QSPI_HYPERFLASH flash_nxp_s32_qspi_hyperflash.c)
+if(CONFIG_FLASH_NXP_S32_QSPI_NOR OR CONFIG_FLASH_NXP_S32_QSPI_HYPERFLASH)
+zephyr_library_sources(flash_nxp_s32_qspi.c)
+zephyr_library_include_directories(
```
cmake indent is 2 spaces

manuargue commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+  drv-strength:
+    type: string
+    enum:
+      - DRV STRENGTH 000 # Typical impedance for 1.8V: 27, typical impedance 3V: 20
```
move the comments to the description, so that they appear in the binding docs

congnguyenhuu replied:
I updated

manuargue commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+  device-id:
+    type: uint8-array
+    required: true
+    description: Device memory ID
```
According to the DS of the hyperflash device connected on this board, it supports [SFDP JEDEC JESD216 serial flash discoverable parameters (SFDP) space](https://www.infineon.com/dgdl/Infineon-S26HS256T_S26HS512T_S26HS01GT_S26HL256T_S26HL512T_S26HL01GT_256Mb_512Mb_1Gb_SEMPER_TM_Flash_HYPERBUS_TM_interface_1-DataSheet-v67_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0ee6a0c96f54). So this binding should be inheriting from jesd216 as well

congnguyenhuu replied:
I removed this property and updated to use "jedec-id"

manuargue commented on drivers/flash/Kconfig.nxp_s32:
```diff
@@ -14,6 +14,18 @@ config FLASH_NXP_S32_QSPI_NOR
+
 if FLASH_NXP_S32_QSPI_NOR
 
 config FLASH_NXP_S32_QSPI_NOR_SFDP_RUNTIME
```
what prevent us to support SFDP runtime configuration for hyperflash?

manuargue commented on boards/nxp/s32z2xxdc2/s32z2xxdc2_s32z270.dtsi:
```diff
@@ -58,3 +58,55 @@
+		end-address = <DT_SIZE_M(512)>;
+	};
+
+	s26hs512t: s26hs512t@0 {
```
According to the board schematics, rev D has a S76HS512TC0BHB010 and rev B a S71KS512C0B connected to QSPI0 A1 port, and both are hyperram and not hyperflash. Please clarify

congnguyenhuu replied:
Maybe there is the difference between the schematic board versions.
The board that I'm using, mounted a hyperflash S26HS512 on board.
I will investigate more about this.
About 'S76HS512TC0BHB010' and 'S71KS512C0B' they are multi-chip package, include both hyperflash chip and hyperram chip. So I believe that the support can work on them although I not have device to test with them.

congnguyenhuu replied:
As my investigation, the original schematic design uses S71KS on S32Z27 DC2 but this part could not be sourced during the pandemic so it is replaced by S26HS on board. So there is newer schematic updated to use S26HS.

manuargue reviewed (changes requested):
is the new driver introduced a hyperbus controller (hyperram & hyperflash), or is it specific only to hyperflash devices?

congnguyenhuu commented:
> is the new driver introduced a hyperbus controller (hyperram & hyperflash), or is it specific only to hyperflash devices?

This driver only introduced for supporting hyperflash devices

congnguyenhuu commented on drivers/flash/flash_nxp_s32_qspi_hyperflash.c:
```diff
@@ -0,0 +1,225 @@
+LOG_MODULE_REGISTER(nxp_s32_qspi_hyperflash, CONFIG_FLASH_LOG_LEVEL);
+
+/* Use the fixed command sets from Qspi_Ip_Hyperflash.c */
+extern Qspi_Ip_InstrOpType Qspi_Ip_HyperflashLutTable[QSPI_IP_HF_LUT_SIZE];
```
This LUT implements a standard ops, it can work with multiple hyperflash memories

congnguyenhuu commented on drivers/flash/flash_nxp_s32_qspi_hyperflash.c:
```diff
@@ -0,0 +1,225 @@
+LOG_MODULE_REGISTER(nxp_s32_qspi_hyperflash, CONFIG_FLASH_LOG_LEVEL);
+
+/* Use the fixed command sets from Qspi_Ip_Hyperflash.c */
+extern Qspi_Ip_InstrOpType Qspi_Ip_HyperflashLutTable[QSPI_IP_HF_LUT_SIZE];
```
This LUT table is fixed definition of HAL RTD and used in HAL source code.
So can not change its name

congnguyenhuu replied:
I updated

congnguyenhuu commented on drivers/flash/flash_nxp_s32_qspi_hyperflash.c:
```diff
@@ -0,0 +1,225 @@
+	int ret = 0;
+
+	/* Used by the HAL to retrieve the internal driver state */
+	data->instance = flash_nxp_s32_qspi_inst_cnt++;
```
I created a function for counting device instance

congnguyenhuu commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+properties:
+  program-page-size:
+    type: int
+    default: 256
```
I updated

congnguyenhuu commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+include: ["nxp,s32-qspi-device.yaml", "soc-nv-flash.yaml"]
+
+properties:
+  program-page-size:
```
I moved it to nxp,s32-qspi-device.yaml

congnguyenhuu commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,79 @@
+
+  secure-region-locked:
+    type: boolean
+    description: Locked secure region
```
I updated

Dat-NguyenDuy commented on dts/bindings/qspi/nxp,s32-qspi-sfp-frad.yaml:
```diff
@@ -0,0 +1,61 @@
+    type: int
+    default: 0
+    description: |
+      The flash memory region start address (around 64 KB boundary), default to 0.
```
- Pls remove default value in the description for all properties (as `default:` feature from dts already describes this)
- Pls also explain why the default value is chosen

Dat-NguyenDuy replied:
If possible, pls add explanation in devicetree bindings, not via PR's comment.
`Pls remove default value in the description for all properties`: what i meant is removing `default to 0` in `The flash memory region start address (around 64 KB boundary), default to 0` (also for all places) because you already use `default:` feature to describe default value

congnguyenhuu replied:
I removed the description mentions about the default value

Dat-NguyenDuy commented on dts/bindings/mtd/nxp,s32-qspi-device.yaml:
```diff
@@ -20,3 +20,15 @@ properties:
       unaligned accesses.
       If not provided, 1 byte alignment will be selected.
+
+  program-page-size:
```
It's look to me this properties should be marked as `required: true`. Pls explain why the default value is chosen

Dat-NguyenDuy replied:
`It's look to me this properties should be marked as required: true`: if this is correct, pls update.
`Pls explain why the default value is chosen`, i meant add explanation in devicetree bindings, not via PR's comment

nordicjm replied:
If a default is set, the description text needs to explain where that default comes from

congnguyenhuu replied:
I update explanation for  the default value

Dat-NguyenDuy reviewed (changes requested):
Some nits in devicetree bindings

gmarull commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,115 @@
+      If it is disable, having access to all memory regions is beneficial during development
+      or debugging phases.
+
+  drv-strength:
+    type: string
+    enum:
+      - DRV STRENGTH 000
+      - DRV STRENGTH 001
+      - DRV STRENGTH 002
+      - DRV STRENGTH 003
+      - DRV STRENGTH 004
+      - DRV STRENGTH 005
+      - DRV STRENGTH 006
+      - DRV STRENGTH 007
+    default: DRV STRENGTH 000
+    description: |
+      Specifies the output driver strength of the device.
+      Typical impedance values for 1.8V:
+        - DRV STRENGTH 000: 27 Ohm
+        - DRV STRENGTH 001: 117 Ohm
+        - DRV STRENGTH 002: 68 Ohm
+        - DRV STRENGTH 003: 45 Ohm
+        - DRV STRENGTH 004: 34 Ohm
+        - DRV STRENGTH 005: 27 Ohm
+        - DRV STRENGTH 006: 24 Ohm
+        - DRV STRENGTH 007: 20 Ohm
+      Typical impedance values for 3V:
+        - DRV STRENGTH 000: 20 Ohm
+        - DRV STRENGTH 001: 71 Ohm
+        - DRV STRENGTH 002: 40 Ohm
+        - DRV STRENGTH 003: 27 Ohm
+        - DRV STRENGTH 004: 20 Ohm
+        - DRV STRENGTH 005: 16 Ohm
+        - DRV STRENGTH 006: 14 Ohm
+        - DRV STRENGTH 007: 12 Ohm
```
this needs to be specified properly, eg, using `drive-strength-ohm = <12>;`.

congnguyenhuu replied:
it is difficult to use directly the ohm value. we need to separate many case.
Because HAL API requires the input argument that is a register value (range from 0 to 7)
Each option corresponds to a typical impedance value as above.
So I'm using a type string property that map define enum of HAL(eg, 'DRV STRENGTH 007' map QSPI_IP_HF_DRV_STRENGTH_007 and corresponds  12 ohm for 3V).
it is more easy than

manuargue replied:
```
enum {
  DRIVE_STRENGTH_20_OHM = QSPI_IP_HF_DRV_STRENGTH_000,
  DRIVE_STRENGTH_27_OHM = QSPI_IP_HF_DRV_STRENGTH_000,
  DRIVE_STRENGTH_117_OHM = QSPI_IP_HF_DRV_STRENGTH_001,
  DRIVE_STRENGTH_71_OHM = QSPI_IP_HF_DRV_STRENGTH_001,
  ...
}

#define DRIVE_STRENGTH(ohm) DRIVE_STRENGTH_ ## ohm ## _OHM
```

congnguyenhuu replied:
This is impossible. So I removed this property.
Because this is an optional setting for flash device, not  mandatory.
We can keep the default reset configuration, not impact to device operation.

manuargue replied:
> This is impossible. So I removed this property.

we just need to add two properties, e.g. `drive-strength-ohm` and `impedance`, and combine them to get the desired HAL enum.

> Because this is an optional setting for flash device, not mandatory.

I'm okay to have the minimal configuration on this pr if it can work with current board supported, it can be extended later on.

> We can keep the default reset configuration (the reset value of register when not setting), not impact to device operation.

where is this default set? it should be documented and I'd also like to see it explicit in the code

congnguyenhuu replied:
okay, I added two properties `drive-strength-ohm` and `vcc-mv` for and handle.

manuargue commented on drivers/flash/flash_nxp_s32_qspi.h:
```diff
@@ -0,0 +1,112 @@
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	struct flash_pages_layout layout;
+#endif
+#if !defined(CONFIG_FLASH_NXP_S32_QSPI_NOR_SFDP_RUNTIME)
```
this currently only apply to NOR flash memories because of the HAL limitation you mentioned. If we want to keep it in the common driver sources, then drop the `_NOR_` and make it depend on `FLASH_NXP_S32_QSPI_NOR`. All SFDP runtime code can be moved to the common driver then.

congnguyenhuu replied:
I updated

manuargue commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,82 @@
+      If it is disable, having access to all memory regions is beneficial during development
+      or debugging phases.
+
+  read-latency:
```
having a second look at this, it seems simpler `read-latency-cycles = <x>` and use an int enum

congnguyenhuu replied:
I updated

manuargue replied:
suffix `-cycles`

congnguyenhuu replied:
I updated

manuargue commented on dts/bindings/qspi/nxp,s32-qspi-sfp-frad.yaml:
```diff
@@ -0,0 +1,62 @@
+  master-domain-acp:
+    type: string
+    enum:
+      - ACP NONE
+      - ACP SECURE
+      - ACP SECURE PRIVILEGED
+      - ACP ALL
+      - ACP PRIVILEGED
```
we can drop the ACP prefix in all enums, already mentioned in the property

congnguyenhuu replied:
I droped

manuargue commented on dts/bindings/qspi/nxp,s32-qspi-sfp-frad.yaml:
```diff
@@ -0,0 +1,62 @@
+  exclusive-access-lock:
+    type: string
+    enum:
+      - EAL DISABLED
+      - EAL NONE
+      - EAL OWNER
```
we can drop the EAL prefix in all enums, already mentioned in the property

congnguyenhuu replied:
I droped

manuargue commented on dts/bindings/qspi/nxp,s32-qspi-sfp-mdad.yaml:
```diff
@@ -0,0 +1,53 @@
+  mask-type:
+    type: string
+    enum:
+      - MASK AND
+      - MASK OR
```
we can drop the MASK prefix in all enums, already mentioned in the property

congnguyenhuu replied:
I droped

manuargue commented on dts/bindings/qspi/nxp,s32-qspi-sfp-mdad.yaml:
```diff
@@ -0,0 +1,53 @@
+    description: |
+      Defines the mask value for the ID-Match comparison.
+      Valid range: 0 - 63. The default corresponds to the
+      reset value of the resgister field.
```
```suggestion
      reset value of the register field.
```

congnguyenhuu replied:
I fixed

congnguyenhuu commented:
Hello @de-nordic, could you visit this PR? thanks.

fabiobaltieri commented on dts/bindings/mtd/nxp,s32-qspi-hyperflash.yaml:
```diff
@@ -0,0 +1,103 @@
+
+  param-sector:
+    type: string
+    enum:
```
oof, yeah, it's valid yaml so the yaml format is perfectly happy with this, there is some basic binding check in one of the compliance runs, I think it makes sense to start expanding on it

congnguyenhuu replied:
This is not DT HALism, this is a key string for reminiscent of setting options. To more nice I updated use underscores instead of spaces.

gmarull replied:
It looks like a HALism because it is intended to map to "register field" (whatever that is), forgetting DT needs to describe things naturally, like `ppw-sectors-address-map = "high";`, or `master-domain-acp-policy = <(XX_PRIVILEGED | XX_SECURE)>;` etc. It seems like some vendors are creating their own DT flavor with a touch of HAL.

congnguyenhuu replied:
I have updated as your suggestion.

manuargue reviewed (dismissed):
all my comments addressed

congnguyenhuu commented:
Hello @de-nordic, could you visit this PR? thanks.

manuargue reviewed (dismissed):
reapproving, my comments addressed

congnguyenhuu commented:
I have just rebased to fix conflict the west manifest. Could you please revisit PR? @nordicjm @manuargue @Dat-NguyenDuy

congnguyenhuu commented:
Hello @de-nordic, could you visit this PR? thanks.

congnguyenhuu committed:
drivers: memc_nxp_s32_qspi: change DT_REG_ADDR to DT_REG_ADDR_RAW

Following the commit f98fde07b30, DT_REG_ADDR now expands with a 'U'
suffix as an unsigned value. However, for compatibility with IS_EQ,
a raw value without any suffix is required. Therefore, this update is
necessary.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu commented on dts/bindings/qspi/nxp,s32-qspi-sfp-frad.yaml:
```diff
@@ -0,0 +1,56 @@
+include: base.yaml
+
+properties:
+  start-address:
+    type: int
+    required: true
+    description: |
+      Specifies the specific flash memory region start address.
+
+  end-address:
+    type: int
+    required: true
+    description: |
+      Specifies the specific flash memory region end address.
```
There is a problem prevent me, the QSPI node has multiple child nodes with different compatible bindings (device flash nodes). In some cases, they all use reg but it is possible dupliacte, which triggers a duplicate reg warning. To avoid this, I decided to use the reg property only for the device flash node.
Do you have any suggestions on how to handle this?@nordicjm

nordicjm replied:
I don't know what you mean? You can have a reg inside of a reg, that's how flash partitions work

congnguyenhuu replied:
use-case:
```
&qspi0 {
	sfp_frad_0: sfp_frad@0 {
		compatible = "nxp,s32-qspi-sfp-frad";
		reg = <0x0 DT_SIZE_M(512)>;
	};

	s26hs512t: s26hs512t@0 {
		compatible = "nxp,s32-qspi-hyperflash";
		reg = <0>;
		status = "okay";

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			storage_partition: partition@0 {
				label = "storage";
				reg = <0x0 0x10000>;
			};
		};
	};
```
the log warning:
![image](https://github.com/user-attachments/assets/ecf79768-8a4a-4e25-9eb1-1d5558b76736)

congnguyenhuu replied:
I found solution for this, move `sfp_frad_0` into a parent node, look like as:
```
&qspi0 {
	sfp {
		compatible = "secure-flash-protection";
		#address-cells = <1>;
		#size-cells = <1>;

	        frad_0: frad@0 {
		        compatible = "nxp,s32-qspi-sfp-frad";
		        reg = <0x0 DT_SIZE_M(512)>;
	        };
        };

	s26hs512t: s26hs512t@0 {
		compatible = "nxp,s32-qspi-hyperflash";
		reg = <0>;
		status = "okay";

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			storage_partition: partition@0 {
				label = "storage";
				reg = <0x0 0x10000>;
			};
		};
	};
```

congnguyenhuu commented on dts/bindings/qspi/nxp,s32-qspi-sfp-frad.yaml:
```diff
@@ -0,0 +1,56 @@
+      Specifies the specific flash memory region end address.
+
+  master-domain-acp-policy:
+    type: int
```
The property allows selecting combinations with multiple configuring values. so I not used enum.
eg. master-domain-acp-policy = <(NXP_S32_QSPI_NON_SECURE | NXP_S32_QSPI_SECURE)>;

congnguyenhuu commented on dts/bindings/qspi/nxp,s32-qspi-sfp-mdad.yaml:
```diff
@@ -0,0 +1,47 @@
+
+properties:
+  secure-attribute:
+    type: int
```
the same above comment

congnguyenhuu commented on dts/bindings/mtd/nxp,s32-qspi-device.yaml:
```diff
@@ -20,3 +20,10 @@ properties:
+    required: true
+    description: |
+      The programming page size of the flash memory device,
+      as specified in the flash memory device datasheet.
```
this property differs with `write-block-size`, this defines maximum of  programming  page buffer size configuration for hardware device flash. To avoid confusing I renamed this property to `max-program-buffer-size`.
As specified configuration in the flash memory device datasheet.
![image](https://github.com/user-attachments/assets/e27ed4f3-9c97-48bb-bd3f-3e1d0eb2a4f2)

nordicjm replied:
Where is the write block size set then, that is required for flash, and erase block size?

congnguyenhuu replied:
write block size is defined by https://github.com/zephyrproject-rtos/zephyr/pull/78073/files#diff-aa5c7e912c23e4bbc1cabe0abb767637bcdaf3f9892233d71bd0841fdbec5c93R13
ease block size is defined by https://github.com/zephyrproject-rtos/zephyr/pull/78073/files#diff-626f2f951a3779218f24c5da9f599d4a6bde9f3fca7e6656cee9531ac1fc1277L46

nordicjm replied:
write block size being a define in code is useless and wrong, there is a dts property, likewise erase block size should be in dts, it's a hardware property (the page size is different, that is what size you want to erase, it is not the erase size)

congnguyenhuu replied:
I see that `write-block-size` and `erase-block-size`, defined in https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mtd/soc-nv-flash.yaml which uses for internal flash.
I refer the other support external flash, I see most them not use these property.
eg. a supported driver that not use these. the write block size is fixed.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/nrf_qspi_nor.c#L1172

congnguyenhuu replied:
Hello @nordicjm, I’ve addressed all your comments. Could you please revisit the PR? I hope you happy with this and can approve my changes so we can merge it soon. Thanks.

congnguyenhuu commented on boards/nxp/mr_canhubk3/mr_canhubk3.dts:
```diff
@@ -344,6 +344,7 @@
 		readoc = "1-4-4";
 		writeoc = "1-4-4";
 		has-32k-erase;
+		program-page-size = <256>;
```
pls see my above comments

congnguyenhuu commented:
I rebased to fix conflict

nordicjm commented on drivers/flash/flash_nxp_s32_qspi.h:
```diff
@@ -0,0 +1,112 @@
+#include "jesd216.h"
+
+#define QSPI_ERASE_VALUE      0xff
+#define QSPI_WRITE_BLOCK_SIZE 1U
```
This needs to come from dts (the write block size). Does your driver switch the commands it sends depending upon what some uses? E.g. if I use flash_write() with 129 bytes it will send 32 commands to write 4 bytes followed by 1 command to write 1 byte or will it just fail at the last part? Likewise what if there is a flash memory used with this that does not support 1 byte writes and is configures for 4 byte writes only?

congnguyenhuu replied:
This driver will use  `memory-alignment` dt properties (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mtd/nxp%2Cs32-qspi-device.yaml#L17) to pad (by adding 0xFF) at the unaligned data parts. E.g with `memory-alignment=2` use `flash_write()` with 129 bytes it will send 130 bytes (last byte is padded if offset is even number). From application pov, `flash_write()` can be used with any size between 1 and the value specified in `max-program-buffer-size` dt properties.

nordicjm replied:
But that is wrong? That's fine if you are using a flash chip, but you get non-flash SPI/QSPI devices (e.g. FRAM) and writing 0xff will write 0xff over data you do not want to touch

congnguyenhuu replied:
This driver supports hyperflash chip (using s26hs512 flash chip on board), writing 0xff not affect the existed data.

nordicjm replied:
And will completely fail with e.g. https://www.infineon.com/cms/en/product/memories/f-ram-ferroelectric-ram/excelon-f-ram/cy15v104qsn-108sxit/

de-nordic replied:
> This driver supports hyperflash chip (using s26hs512 flash chip on board), writing 0xff not affect the existed data.

No hidden writes. Require users to align size and offset to value you are writing. Otherwise you differ from other drivers, at Flash Api level , from other drivers.

congnguyenhuu replied:
I added the property dts `write-block-size` for resolve this.

de-nordic replied:
People that will write code for subsystem to work on any device have to align offset and size to write-block-size anyway, so by not requiring that. So if somebody delivers something for your device with assumption that data will be padded, will have to fix that when testing code against other device.

congnguyenhuu replied:
@de-nordic @nordicjm i believe your concern for this topic has been addressed (by updating the PR to use `write-block-size`), could both of you help me to take another look.

nordicjm replied:
> @de-nordic @nordicjm i believe your concern for this topic has been addressed (by updating the PR to use `write-block-size`), could both of you help me to take another look.

You have added a write block size of 1, so does that mean then when I call this function, your driver **writes exactly one byte with no padding or 0xff's what so ever**?

congnguyenhuu replied:
you are saying [this write block size](https://github.com/zephyrproject-rtos/zephyr/pull/78073/files#diff-49b9c21ea3c86864ed5469210d04a68afb14267f2b59ff4f5711648899c94d7fR348). this is write block size for a serial NOR flash device. it allows to write exactly 1 byte without padding.

nordicjm replied:
I don't mean for a serial NOR flash device, I mean the actual command that comes out the chip, does it send a command to the other device consisting of writing 1 byte and 1 byte only without **any** padding or 0xff's or does it send a command with additional data/padding?

congnguyenhuu replied:
After I changed to use `write-block-size`, driver sends command with only data without padding.

congnguyenhuu replied:
After I changed to use `write-block-size`, driver send command with only data without padding.

nordicjm replied:
Thanks

congnguyenhuu commented:
Hello @nordicjm, after some consideration, I added the `write-block-size` property to the DTS. I noticed it relates to the existing `memory-alignment` property. Therefore, I decided to remove `memory-alignment` and can use` write-block-size` instead to calculate padding for unaligned accesses.

congnguyenhuu commented:
Hello @nordicjm, Do you have any comments? If not, could you please approve my changes? I hope this PR gets merged soon. thanks.

de-nordic commented on dts/bindings/mtd/nxp,s32-qspi-device.yaml:
```diff
@@ -14,9 +14,16 @@ properties:
-      Memory alignment in bytes, used to calculate padding when performing
-      unaligned accesses.
-      If not provided, 1 byte alignment will be selected.
+      The number of bytes used in write operations, it also used to calculate padding when
```
Still mentioning padding.

congnguyenhuu replied:
I updated

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,271 @@
+{
+	Qspi_Ip_MemoryConfigType *memory_cfg = get_memory_config(dev);
+
+	return ((offset >= 0) && (offset < memory_cfg->memSize) &&
```
if size and offset get large enough you  can get into situation where they overflow and you and up having true in line 22, even though parameters are out of range.
More proper would be to do:
```
return (offset >=0 && offset < memory_cfg->memSize && ((memory_cfg->mem_size - offset) >= size)
```

congnguyenhuu replied:
I see the condition in line 22 `((size + offset) <= memory_cfg->memSize))` that get the same result as your suggestion condition `((memory_cfg->mem_size - offset) >= size)`

congnguyenhuu replied:
I updated

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	if (!area_is_subregion(dev, offset, size)) {
+		return -ENODEV;
```
I am quite sure other drivers return -EINVAL here.

de-nordic replied:
There is no check of offset nor size for alignment to write-block-size set in DTS, I know that device permits 1, but if you set it to something else and let to be ignored, this will allow different subsystems write/read with different alignment and produce data that they may not be able mutually read from the same device.

congnguyenhuu replied:
I addressed comment

congnguyenhuu replied:
I addressed comment. could you please revisit? @de-nordic

congnguyenhuu commented:
Hello @de-nordic, could you revisit this PR? thanks.

congnguyenhuu commented:
I rebased to fix conflict file west manifest

congnguyenhuu commented:
Hello @dlech, @manuargue, @de-nordic. Is there any chance that this PR can be merged before the freeze?

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,274 @@
+	if (status != STATUS_QSPI_IP_SUCCESS) {
+		LOG_ERR("Failed to read memory status (%d)", status);
+		ret = -EIO;
+	} else if (timeout <= 0) {
```
timeout will never be negative, it is uint32_t

congnguyenhuu replied:
I updated this condition

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,274 @@
+	uint32_t timeout = 0xFFFFFF;
+	int ret = 0;
+
+	do {
+		status = Qspi_Ip_GetMemoryStatus(data->instance);
+		timeout--;
+	} while ((status == STATUS_QSPI_IP_BUSY) && (timeout > 0));
```
You really want to busy loop here? Is there any idling in Qspi_Ip_GetMemoryStatus ?

congnguyenhuu replied:
there is not waiting timeout in `Qspi_Ip_GetMemoryStatus ` so it is needed.

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,274 @@
+	size_t max_write = (size_t)MIN(QSPI_IP_MAX_WRITE_SIZE, memory_cfg->pageSize);
+	size_t len;
+	int ret = 0;
+
```
is size is 0 there is a lot happening, without write, that could be just bypassed.

congnguyenhuu replied:
I updated return -EINVAL

congnguyenhuu replied:
I updated return -EINVAL when size = 0

de-nordic commented on drivers/flash/flash_nxp_s32_qspi.c:
```diff
@@ -0,0 +1,274 @@
+	Qspi_Ip_StatusType status;
+	size_t erase_size;
+	int ret = 0;
+
```
Same here, why not just exit when size is 0?

congnguyenhuu replied:
I updated return -EINVAL

de-nordic reviewed (dismissed):
Minor comments left.

Configuration for storage samples, look ok.

mmahadevan108 commented:
@congnguyenhuu, kindly revisit this PR.

congnguyenhuu commented:
hello @manuargue @Dat-NguyenDuy and @de-nordic, I have rebased and addressed the some minor comments from de-nordic.
Could you please help me revisit. thanks.

congnguyenhuu commented:
Hello @de-nordic, Could you please help me revisit. thanks.

danieldegrasse commented:
Something to be aware of- longer term, we should likely be implementing this peripheral using the MSPI API https://docs.zephyrproject.org/latest/hardware/peripherals/mspi.html, right? I don't believe we should block this from moving forwards based on that but it is something to be aware of- long term we should have one flash driver that probes SFDP data from the attached flash device using the MSPI API to interact with the controller, IMO.

manuargue commented:
> Something to be aware of- longer term, we should likely be implementing this peripheral using the MSPI API https://docs.zephyrproject.org/latest/hardware/peripherals/mspi.html, right? 

I wasn't aware of this relatively new device class, definitely seems to be the right choice, thanks for the heads up. We'll look at that.

congnguyenhuu commented:
I rebased to fix conflict west manifest

Dat-NguyenDuy commented:
Hi @de-nordic, could you please re-approve this PR so it can be merged. Thanks

de-nordic reviewed (dismissed):
Tests and samples look OK.

@danieldegrasse @dleach02 Please review changes to NXP drivers.

danieldegrasse commented on boards/nxp/s32z2xxdc2/s32z2xxdc2_s32z270.dtsi:
```diff
@@ -58,3 +59,67 @@
+	status = "okay";
+
+	sfp {
+		compatible = "secure-flash-protection";
```
Is this compatible needed? I don't see it defined in tree.

congnguyenhuu replied:
It is needed for building this device tree node.

danieldegrasse replied:
I'm not sure I follow- if we don't have a compatible here, does the support no longer work?

congnguyenhuu replied:
I previously added the parent node 'sfp' to address a reviewer's [comment](https://github.com/zephyrproject-rtos/zephyr/pull/78073#discussion_r1798966791). Its binding is not defined but it is required for this node. By adding the 'secure-flash-protection' compatibility, we can differentiate it from other nodes. Without this, Zephyr would build this node as 'nxp,s32-qspi-hyperflash', which requires non-existent properties and causes build failures.

congnguyenhuu replied:
I updated to remove this compatible

danieldegrasse commented on west.yml:
```diff
@@ -199,7 +199,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: c42d70dba969de09c0e43d3b26a3ffcb015f6f33
+      revision: pull/434/head
```
Why is HAL update made in the commit adding board support? Are the HAL changes only needed once the board is enabled?

congnguyenhuu replied:
I think this HAL update should be included in this commit since it is being enabled the flash node. Without the correct HAL update, the build will fail in this commit.

congnguyenhuu replied:
I updated to move this into the proper commit

congnguyenhuu commented:
I rebased to fix conflict west manifest

congnguyenhuu commented:
Hello @danieldegrasse, Could you help me to review these changes. Thanks.

de-nordic commented:
@congnguyenhuu please fix the ci failures.

congnguyenhuu commented:
Hello, I rebased to fix the CI check failure, but the current bsim test check is failing based on the latest code. Therefore, I logged issue #83388 to fix.

congnguyenhuu commented:
Hello @de-nordic, the CI failure is fixed, could you please visit. Thanks.

congnguyenhuu commented:
Hello @de-nordic, the CI failure is fixed, could you please revisit. Thanks.

congnguyenhuu commented:
Hello @de-nordic, Could you please review the changes again? We need your approval as the assignee to merge the HAL PR first, so we can update the west manifest file before merging this PR. Thanks.

manuargue commented:
@dleach02 can you please help us move forward with this merge? this PR have been approved a few times already and the HAL pr is also approved. Thanks

de-nordic reviewed (dismissed):
Storage related tests and samples look ok.
@dleach02 NXP specific stuff is on your people to review.

mmahadevan108 reviewed (changes requested):
Needs to be sequenced behind https://github.com/zephyrproject-rtos/zephyr/pull/81012
@congnguyenhuu  I merged the HAL side. After the gating PR merges, then we can update the HAL SHA and I'll clear my CR

congnguyenhuu commented:
> Needs to be sequenced behind #81012 @congnguyenhuu I merged the HAL side. After the gating PR merges, then we can update the HAL SHA and I'll clear my CR

I updated the HAL SHA

mmahadevan108 commented:
@de-nordic , could you please re-approve. Lost your approval when updating the HAL SHA

congnguyenhuu commented:
nm
